rfcs/0002-adding-minion-id-to-fileserver-calls.md

```diff
@@ -0,0 +1,44 @@
+- Feature Name: adding_minion_id_to_fileserver_calls
+- Start Date: 2018-09-13
+- RFC PR:
+- Salt Issue:
+
+# Summary
+[summary]: #summary
+
+Add minion id to all fileserver events sent to request server and optionally sign the payload as well.
+
+# Motivation
+[motivation]: #motivation
+
+Currently there is no possibility to determine on master which minion requested particular fileserver operation (file list, file request, etc.), which makes it impossible to implement any minion-based permissions level onto fileserver.
+
+Implementing this RFC would make it possible to implement custom fileservers that support various layers of security, allowing to restrict access for particular minions to environments/files and achieving some level of fileserver multitenancy.
+
+# Design
+[design]: #detailed-design
+
+Proposed design is to add `id` key to all fileserver-related payloads in `RemoteClient`, optionally allowing to also sign the messages similarly as it's done via [minion_sign_messages](https://github.com/saltstack/salt/blob/cb172352340bf8da34cd128dbddf1abfc4995bb5/salt/minion.py#L1380) (either reusing same configuration option or introducing new one).
```

Comment:
Well If I understand the master-minion communication, there is none contextual information when handling any message from minion on master level (https://github.com/saltstack/salt/blob/develop/salt/master.py#L1035, which basically just executes any function from https://github.com/saltstack/salt/blob/develop/salt/master.py#L1141, ignoring any methods starting with __ though). So you are left with authenticated message - you know it's *some* minion, you just don't know which one.

Current methods are solving this with just passing `tok` in the payload and verifying it with https://github.com/saltstack/salt/blob/develop/salt/master.py#L1189, which is used for example for mines, pillars obtaining, etc. That is however just `salt` string encrypted with minion private key. When you switch on the option to encrypt return events from minion (`minion_sign_messages` in https://github.com/saltstack/salt/blob/develop/salt/minion.py#L1383, so basically any fire_master events, which is processed in https://github.com/saltstack/salt/blob/develop/salt/master.py#L1616 on master side), master doesn't checks the `salt` token (never did for these events), but verifies the whole message - so you know the message was really crafted by somebody with access to the private key - while when using `tok` you just know it got sent by somebody having the encrypted `salt` string). However neither of those things are sent by minion for fileserver requests - so you have no possibility to actually determine who requested the file - you just know it was *some* minion.

So the first obvious option is to reuse the same thing also for the fileserver requests - the events should be pretty small (the upload functionality with `file_recv` is handled separately via https://github.com/saltstack/salt/blob/develop/salt/master.py#L1431 and already contains `tok`), so verifying the whole event would be better security-wise, but using `tok` would be sufficient and could be speeded up in the future (as you know what the `tok` should look like - it's not unique for specific request).

Also the second and way better solution would be to just have the contextual information about the minion when handling any master function, but that isn't probably that easy as it sounds, because it would be already implemented (just guessing, I can dig into it more to find a way how to do it - if it can even be done).

Reply:
Yes, I understand the problem that it was _some_ minion. I am thinking just how to avoid revalidation. We can use token to determine what _minion_ and use request code (?) or description to determine what _request type_. So we still do not need to compute things.

I am not sure we need here security within security. We already connected the minion. We've got root on it, we running Master as non-root. So it is all decided already anyway. The only thing left is to know who is who and why comes. For that I so far cannot understand why would we need to sign every message and verify them on master.

But the rest to me looks fine, I like it.